File: src/main/java/toast/client/commands/cmds/Macro.java
package toast.client.commands.cmds;

import toast.client.ToastClient;
import toast.client.commands.Command;
import toast.client.utils.KeyUtil;
import toast.client.utils.Logger;

import static toast.client.ToastClient.CONFIG_MANAGER;

public class Macro extends Command {
    public Macro() {
        super("Macro", ToastClient.cmdPrefix + "macro [add/remove/list] <key> <command/message>", "Allows you to bind a message to a key", false, "macros", "macro");
    }

    public void run(String[] args) {
        if (args.length == 0) {
            Logger.message("Missing arguments!", Logger.ERR, false);
            return;
        }
        switch (args[0]) {
            case "add":
                if (args.length >= 3) {
                    if (KeyUtil.isNumeric(args[1])) {
                        try {
                            System.out.println(Integer.parseInt(args[1]));
                            CONFIG_MANAGER.writeMacros(args[2], Integer.parseInt(args[1]));
                        } catch (NumberFormatException nfe) {
                            Logger.message("Failed to add macro.", Logger.ERR, false);
                            System.out.println("Failed");
                        }
                    }
                    return;
                }
                Logger.message("Missing arguments!", Logger.ERR, false);
            case "remove":
                if (args[1] == null) {
                    Logger.message("Missing arguments!", Logger.ERR, false);
                    return;
                }
                if (KeyUtil.isNumeric(args[1])) {
                    try {
                        System.out.println(Integer.parseInt(args[1]));
                        CONFIG_MANAGER.macros.forEach((command, key) -> {
                            if (key == Integer.parseInt(args[1])) {
                                CONFIG_MANAGER.macros.remove(command, key);
                                CONFIG_MANAGER.loadMacros();
                            }
                        });

                    } catch (NumberFormatException nfe) {
                        Logger.message("Failed to remove macro.", Logger.ERR, false);
                        System.out.println("Failed");
                    }
                }
            case "list":
                final String[] message = {"Key | Message\n"};
                CONFIG_MANAGER.macros.forEach((command, key) -> message[0] += (key + " | " + command));
                Logger.message(message[0], Logger.INFO, false);
            default:
                Logger.message("Could not parse command.", Logger.ERR, false);
        }
    }

}

File: src/main/java/toast/client/commands/cmds/Macro.kt
package toast.client.commands.cmds

import org.lwjgl.glfw.GLFW
import toast.client.ToastClient
import toast.client.commands.Command
import toast.client.utils.KeyUtil
import toast.client.utils.MessageUtil
import java.util.*

/**
 * Command to add, remove and list existing macros
 */
class Macro : Command("Macro", """${ToastClient.cmdPrefix}macro [add/remove/list] <key> <command/message>""", "Allows you to bind a message to a key", false, "macros", "macro") {
    override fun run(args: Array<String>) {
        if (args.isEmpty()) {
            MessageUtil.sendMessage("Missing arguments!", MessageUtil.Color.RED)
            return
        }
        when (args[0]) {
            "add" -> {
                if (args.size >= 3) {
                    val keyCode = KeyUtil.getKeyCode(args[1])
                    if (keyCode == -1) {
                        MessageUtil.sendMessage("No Key by the name of ${args[1]} was found.", MessageUtil.Color.RED)
                        return
                    }
                    try {
                        var command = args[2]
                        if (args.size > 3) {
                            for (x in 3 until args.size) {
                                command += " ${args[x]}"
                            }
                        }
                        ToastClient.CONFIG_MANAGER.loadMacros()
                        ToastClient.CONFIG_MANAGER.addMacro(command, keyCode)
                        MessageUtil.sendMessage("Added macro: ${args[1]} | $command", MessageUtil.Color.GREEN)
                    } catch (nfe: NumberFormatException) {
                        MessageUtil.sendMessage("Failed to add macro.", MessageUtil.Color.RED)
                    }
                    return
                }
                MessageUtil.sendMessage("Missing arguments!", MessageUtil.Color.RED)
            }
            "remove" -> {
                val keyCode = KeyUtil.getKeyCode(args[1])
                if (keyCode == -1) {
                    MessageUtil.sendMessage("No Key by the name of ${args[1]} was found.", MessageUtil.Color.RED)
                    return
                }
                try {
                    println(keyCode)
                    val macroIter = (ToastClient.CONFIG_MANAGER.getMacros() ?: return).iterator()
                    while (macroIter.hasNext()) {
                        val next = macroIter.next()
                        if (next.value == keyCode) {
                            ToastClient.CONFIG_MANAGER.loadMacros()
                            (ToastClient.CONFIG_MANAGER.getMacros() ?: continue).remove(next.key)
                            ToastClient.CONFIG_MANAGER.writeMacros()
                            MessageUtil.sendMessage("Removed macro: ${args[1]} | ${next.key}", MessageUtil.Color.GREEN)
                            return
                        }
                    }
                    MessageUtil.sendMessage("No macro bound to ${args[1]} was found.", MessageUtil.Color.RED)
                } catch (t: Throwable) {
                    MessageUtil.sendMessage("Failed to remove macro.", MessageUtil.Color.RED)
                }
            }
            "list" -> {
                val messages = ArrayList<String>()
                messages.add("KeyNum | KeyName | Message")
                (ToastClient.CONFIG_MANAGER.getMacros() ?: return).forEach { (command, key) ->
                    messages.add(key.toString() + " | " + GLFW.glfwGetKeyName(key
                            ?: return@forEach, GLFW.glfwGetKeyScancode(key)) + " | " + command)
                }
                for (message in messages) {
                    MessageUtil.sendMessage(message, MessageUtil.Color.GRAY)
                }
            }
            else -> MessageUtil.sendMessage("Could not parse command.", MessageUtil.Color.RED)
        }
    }
}